0000-unified-machine-word-trait.md

```diff
@@ -0,0 +1,196 @@
+- Feature Name: unified_machine_word_arithmetic
+- Start Date: 2016-07-13
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Unify functionality peculiar to `i8`…`i64` and `u8`…`u64` in a trait containing the
+family of `overflowing`/`checked`/`wrapping`/`saturating` variants of arithmetic operations,
+as well as a few new ones.
+
+# Motivation
+[motivation]: #motivation
+
+The current design has the following functions replicated over eight different types:
+
+- `checked_add`
+- `checked_div`
+- `checked_mul`
+- `checked_neg`
+- `checked_rem`
+- `checked_shl`
+- `checked_shr`
+- `checked_sub`
+- `count_ones`
+- `count_zeros`
+- `leading_zeros`
+- `overflowing_add`
+- `overflowing_div`
+- `overflowing_mul`
+- `overflowing_neg`
+- `overflowing_rem`
+- `overflowing_shl`
+- `overflowing_shr`
+- `overflowing_sub`
+- `rotate_left`
+- `rotate_right`
+- `saturating_add`
+- `saturating_mul`
+- `saturating_sub`
+- `trailing_zeros`
+- `wrapping_add`
+- `wrapping_div`
+- `wrapping_mul`
+- `wrapping_neg`
+- `wrapping_rem`
+- `wrapping_shl`
+- `wrapping_shr`
+- `wrapping_sub`
+
+Further functions that are replicated include:
+
+- `from_be`
+- `from_le`
+- `swap_bytes`
+- `to_be`
+- `to_le`
+
+But some of those should get their own trait dealing with endianness, which is not the
+focus of this RFC.
+
+Generalizing the arithmetic functionality into a trait, will have the following benefits:
+
+- Enable generic code working with arithmetic overflow — a clunky process at present.
+- Pave the way for general-case software implementations of large integer sizes on platforms
+    without support.
+- Make it easy to implement `i128`/`u128` (and larger types) when compiler/hardware support permits, letting
+    generic code predating the addition of the larger types work with it.
+- Generic implementation of `std::ops::*` traits with the usual panic-on-overflow semantics, using the
+    `checked_*` functions.
+
+Philosophically, one might argue that a solid trait for machine-arithmetic might pave the way
+for saner arithmetic hierarchies: machine-level arithmetic is seldom easily expressible in terms of
+abstract algebra; indeed this RFC may be said to enforce a separation of concerns in library design.
+
+# Detailed design
+[design]: #detailed-design
+
+I propose the following trait to be added to the standard library where convenient:
+
+    trait MachineArith : Copy + Eq {
+      fn abs(self) -> Self;
```

Comment:
I'm concerned about including `abs` here, seeing as we might want to use this in an algebraic trait hierarchy. What is your reasoning for including it here?